File: src/main/java/ee/carlrobert/codegpt/util/DownloadingUtil.java
package ee.carlrobert.codegpt.util;

import static java.lang.String.format;

import ee.carlrobert.codegpt.util.file.FileUtil;

public class DownloadingUtil {

  private DownloadingUtil() {
  }

  private static final int BYTES_IN_MB = 1024 * 1024;

  public static String getFormattedDownloadProgress(long startTime, long fileSize, long bytesRead) {
    long timeElapsed = System.currentTimeMillis() - startTime;

    double speed = ((double) bytesRead / timeElapsed) * 1000 / BYTES_IN_MB;
    double percent = (double) bytesRead / fileSize * 100;
    double downloadedMB = (double) bytesRead / BYTES_IN_MB;
    double totalMB = (double) fileSize / BYTES_IN_MB;
    double remainingMB = totalMB - downloadedMB;

    return format(
        "%s of %s (%.2f%%), Speed: %.2f MB/sec, Time left: %s",
        FileUtil.convertFileSize((long) downloadedMB * BYTES_IN_MB),
        FileUtil.convertFileSize((long) totalMB * BYTES_IN_MB),
        percent,
        speed,
        getTimeLeftFormattedString(speed, remainingMB));
  }

  private static String getTimeLeftFormattedString(double speed, double remainingMB) {
    double timeLeftSec = speed > 0 ? remainingMB / speed : 0;
    long hours = (long) (timeLeftSec / 3600);
    long minutes = (long) ((timeLeftSec % 3600) / 60);
    long seconds = (long) (timeLeftSec % 60);

    return format("%02d:%02d:%02d", hours, minutes, seconds);
  }
}

File: src/main/kotlin/ee/carlrobert/codegpt/util/DownloadingUtil.kt
package ee.carlrobert.codegpt.util

import ee.carlrobert.codegpt.util.file.FileUtil.convertFileSize

object DownloadingUtil {
  private const val BYTES_IN_MB = 1024 * 1024

  fun getFormattedDownloadProgress(
    fileNumber: Int, fileCount: Int, startTime: Long,
    fileSize: Long, bytesRead: Long
  ): String {
    val timeElapsed = System.currentTimeMillis() - startTime

    val speed = (bytesRead.toDouble() / timeElapsed) * 1000 / BYTES_IN_MB
    val percent = bytesRead.toDouble() / fileSize * 100
    val downloadedMB = bytesRead.toDouble() / BYTES_IN_MB
    val totalMB = fileSize.toDouble() / BYTES_IN_MB
    val remainingMB = totalMB - downloadedMB

    return String.format(
      "File %d/%d: %s of %s (%.2f%%), Speed: %.2f MB/sec, Time left: %s",
      fileNumber,
      fileCount,
      convertFileSize(downloadedMB.toLong() * BYTES_IN_MB),
      convertFileSize(totalMB.toLong() * BYTES_IN_MB),
      percent,
      speed,
      getTimeLeftFormattedString(speed, remainingMB)
    )
  }

  private fun getTimeLeftFormattedString(speed: Double, remainingMB: Double): String {
    val timeLeftSec = if (speed > 0) remainingMB / speed else 0.0
    val hours = (timeLeftSec / 3600).toLong()
    val minutes = ((timeLeftSec % 3600) / 60).toLong()
    val seconds = (timeLeftSec % 60).toLong()

    return String.format("%02d:%02d:%02d", hours, minutes, seconds)
  }
}
